airflow/dag_processing/bundles/git.py

```diff
@@ -120,9 +154,16 @@ def _has_version(repo: Repo, version: str) -> bool:
         except BadName:
             return False
 
+    def _refresh(self):
+        self.bare_repo.remotes.origin.fetch("+refs/heads/*:refs/heads/*")
+        self.repo.remotes.origin.pull()
+
     def refresh(self) -> None:
         if self.version:
             raise AirflowException("Refreshing a specific version is not supported")
 
-        self.bare_repo.remotes.origin.fetch("+refs/heads/*:refs/heads/*")
-        self.repo.remotes.origin.pull()
+        if self.ssh_hook:
+            with self.ssh_hook.get_conn():
```

Comment:
Doesn't this create a session to the (a?) remote ssh host [here](https://github.com/apache/airflow/blob/351ac28c7c4dde3a611e4260bbb0f01c7618f261/providers/src/airflow/providers/ssh/hooks/ssh.py#L339-L340)? Not sure we want that. Not even sure where this would connect, since the remote host likely wouldn't match the `repo_url`.

Unless we can somehow tell `git` to use that socket/session?

Reply:
From what I experimented, once the ssh connection is established, it will be used by git to clone the repo. The remote host is where the ssh will establish the connection and not the repo url. While testing in breeze, I set localhost as the remote host and cloned a private repo. Also, in the breeze ssh config, I put it to use private key:
```
Host * 
 IdentityFile ~/.ssh/id_ed25519
```

Maybe @potiuk can help here.

Reply:
I have not paid too much attention to those changes so far but happy to help. Not sure exactly what is the problem, but if I understand correctly the question is about the `repo_url` and `host` established in SSHHook via the SSH conn.

Yeah - generally `git` can use SSH protocol (And I think it should be generally default for most "production" setup configurations. 

https://git-scm.com/book/ms/v2/Git-on-the-Server-The-Protocols

And in this case indeed we need to somehow make repo_url and SSHHook host match. because if they don't then indeed `SSHHook.get_conn()` uses the host from SSH conn to establish connection and they connection is reused later.

I believe (from what I see and earlier comments) we are not trying to use existing ssh_conn_id - which I find a bit strange, because Connection (and/or secrets) is where we can securely store authentication information, so i think we **should** pass conn_id - to retrieve the authentication information from the Connection.

I can imagine - for example that:

```
Connection(
    login = "git"
    host = "github.com"
    extra = {... private_key etc ..)
``

Where repo_url = :

```
git@github.com:apache/airflow.git
```

There is a bit of a problem here indeed because the question is how to make host part of the repo_url and connection host match.  We don't have really a place to store the "repo" portion of the github URL in SSH Connection, so in the current setuo we have a bit of a strange mixup between connection and repo_url where part of the connection definition is in the hook and part in the repo_url.

IMHO - the cleanest solution would be to create a "GItHook" and "Git connection" extending from SSH Hook. We could add "repository" extra to "GitConnection" on top of what "SSH Connection" already has - and use it exclusively for git connectivity. In this case "repo_url" and "ssh_conn_id" would be mutually exclusive. This would avoid a lot of confusion (we have very similar case where we mixed configuration of emails sent by Airflow - where part of the email configuration is taken from the email conn id and part from the configuraiton and it ain't pretty.

Reply:
Yeah, I worked towards a separate hook earlier but later changed to using the ssh conn_id. I will revisit it. I think the issue I had was populating the connection type field with the new hook since the hook is not yet in the providers package.

Reply:
Pushed a change to use githook. Not manually tested for now

Reply:
> Doesn't this create a session to the (a?) remote ssh host [here](https://github.com/apache/airflow/blob/351ac28c7c4dde3a611e4260bbb0f01c7618f261/providers/src/airflow/providers/ssh/hooks/ssh.py#L339-L340)? Not sure we want that. Not even sure where this would connect, since the remote host likely wouldn't match the repo_url.

I have reworked this to use a separate connection derived from SSHHook. This still has to connect to the host machine(not github.com, for example) but the host where the Airflow is running. When connected to the host, it will now use the provided ssh key to authenticate to github and clone the repository. This connection is closed after cloning the repo because of the use of the context manager when establishing the connection.

Reply:
This looks good. Would be nice to add fields for the extras -> json to edit extras is not too convenient.